### ISSUE TITLE
Is retracing inevitable when working with variously-sized inputs?

### ISSUE THREAD

**Issue by cranedroesch** [tensorflow/tensorflow#38002]
I'm beginning to develop some models for token classification with varying-length documents, and it's looking like using variable-length inputs with one-document batches is likely going to give better-results than padding, given the extreme heterogeneity in document lengths.

Here's a simplified version of my model:

```
inp = Input(shape=(None, input_dims))
lay = Conv1D(filters=100, kernel_size=5, padding='same', activation='relu')(inp)
outlayers = [Conv1D(filters=3, kernel_size=5, activation="softmax", padding='same', name=i)(lay) for i in out_varnames]
model = Model(inp, outlayers)
```

The inputs vary in size from `(100,input_dims)` to `(8000,input_dims)`, hence the choice to use variable-length inputs.  

My training function is fairly generic:
```
@tf.function
def train(x, y):
    with tf.GradientTape() as g:
        pred = model(x)
        loss = loss_object(y, pred)
        gradients = g.gradient(loss, model.trainable_variables)
        optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```

When I loop the documents through the train function, I get the following warning:

> WARNING:tensorflow:5 out of the last 5 calls to <function train at 0x7f954a804d40> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/beta/tutorials/eager/tf_function#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.

Reading and github issues elsewhere, I get the impression that retracing is caused by the fact that the varying input lengths require tensorflow to rejigger some stuff under the hood.  

My questions:  
(1) is this accurate?  Is retracing caused by the variably-sized inputs?
(2) is there a way to avoid retracing when working with variably-sized inputs?
(3) what are some ways to optimize performance when working with variably-sized inputs?

**Comment by Saduf2019** [comment]
@cranedroesch 
could you please share the tensor flow version on which the error is faced, along with a simple standalone code for us to replicate the issue

**Comment by asquare92** [comment]
TF = 2.0.0
Python = 3.7.6
CUDA = 10.1.243
cuDNN = 7.6.4.38

I am also having the somewhat same problem as @cranedroesch, but my input-size is not variable. 
I am using a custom training loop with GradientTape to train the model. Everything is working as expected but instead of retracing. A new graph is built on every new training step and that leads to memory leak and my CPU's RAM is accumulating memory with every epoch(and also with every step_per_epoch). Following is my generic training script. 

```
model = YoloV3(yolo_size, training=True, classes=num_classes)
optimizer = tf.keras.optimizers.Adam(lr=learning_rate)
loss = [YoloLoss(anchors[mask], classes=num_classes)
            for mask in anchor_masks]

#both training and validation datasets are python generator.
train_dataset()#returns tf.Tensor( batch of [3-D Images], shape=(1, 416, 416, 3), dtype=float32)
val_dataset()#returns tuple(<tf.Tensor: id=1809, shape=(1, 13, 13, 3, 6), dtype=float32, numpy=array of labels , dtype=float32)>, <tf.Tensor: id=1810, shape=(1, 26, 26, 3, 6), dtype=float32, numpy=array of labels , dtype=float32)>,<tf.Tensor: id=1811, shape=(1, 52, 52, 3, 6), dtype=float32, numpy=array of labels , dtype=float32))>)
#val_dataset is actually a tuple of three eager tensors e.g tuple(has three eager tensors)

avg_loss = tf.keras.metrics.Mean('loss', dtype=tf.float32)
avg_val_loss = tf.keras.metrics.Mean('val_loss', dtype=tf.float32)

# train graph
# also tried with def train(model, optimizer, loss, avg_loss, images, labels):
#as mentioned here https://www.tensorflow.org/api_docs/python/tf/function that you can avoid retracing if you use tensors as the input parameters to the graph and by using input_signature. So also used 
# @tf.function (input_signature=[tf.TensorSpec(shape=[None, None, None, None], dtype=tf.float32),
#                               (tf.TensorSpec(shape=[None, None, None, None], dtype=tf.float32),
#                                 tf.TensorSpec(shape=[None, None, None, None], dtype=tf.float32),
#                                 tf.TensorSpec(shape=[None, None, None, None], dtype=tf.float32)
#                                )])
But it still got the RAM acuumulation issue. 

@tf.function
def train(images, labels):
    with tf.GradientTape() as tape:
        outputs = model(images, training=True)
        regularization_loss = tf.reduce_sum(model.losses)
        pred_loss = []
        for output, label, loss_fn in zip(outputs, labels, loss):
            pred_loss.append(loss_fn(label, output))
        total_loss = tf.reduce_sum(pred_loss) + regularization_loss
    grads = tape.gradient(total_loss, model.trainable_variables)
    optimizer.apply_gradients(zip(grads, model.trainable_variables))
    avg_loss.update_state(total_loss)


# val graph
@tf.function
# def val(model, loss, avg_val_loss, images, labels):
def val(images, labels):
    # print('val is called')
    outputs = model(images, training=False)
    regularization_loss = tf.reduce_sum(model.losses)
    pred_loss = []
    for output, label, loss_fn in zip(outputs, labels, loss):
        pred_loss.append(loss_fn(label, output))
    total_loss = tf.reduce_sum(pred_loss) + regularization_loss
    avg_val_loss.update_state(total_loss)

#Graph retracing and RAM accumulation issue with the following training loop
# training loop: Also updating the progress bar and writing tensorboard summary in the loop that is not shown here.

for epoch in range(epochs):
    avg_loss.reset_states()
    avg_val_loss.reset_states()

    for train_images, train_labels in train_dataset:
        if train_steps_per_epoch != steps_per_epochs:
            train(train_images, train_labels)
            train_loss_value = [('loss', avg_loss.result().numpy())]
            train_steps_per_epoch += 1
            if train_steps_per_epoch == steps_per_epochs:
                break
        else:
            break

    val_step_train = train_steps_per_epoch
    train_steps_per_epoch = 0

    for val_images, val_labels in val_dataset:
        if val_steps_per_epoch != validation_steps:
            val(val_images, val_labels)
            val_loss_value = [('val_loss', avg_val_loss.result().numpy())]
            val_steps_per_epoch += 1
            if val_steps_per_epoch == validation_steps:
                break
        else:
            break
    val_steps_per_epoch = 0
    model.save_weights()

#The only that is working for me currently is resetting the default graph on every step_per_epoch. Although this, I think, is not the solution but the workaround. Following is the training loop that does not accumulate the CPU's RAM on every training step.

for epoch in range(epochs):
    avg_loss.reset_states()
    avg_val_loss.reset_states()

    for train_images, train_labels in train_dataset:
        tf.compat.v1.reset_default_graph()
        if train_steps_per_epoch != steps_per_epochs:
            train(train_images, train_labels)
            train_loss_value = [('loss', avg_loss.result().numpy())]
            train_steps_per_epoch += 1
            if train_steps_per_epoch == steps_per_epochs:
                break
        else:
            break

    val_step_train = train_steps_per_epoch
    train_steps_per_epoch = 0

    for val_images, val_labels in val_dataset:
        tf.compat.v1.reset_default_graph()
        if val_steps_per_epoch != validation_steps:
            val(val_images, val_labels)
            val_loss_value = [('val_loss', avg_val_loss.result().numpy())]
            val_steps_per_epoch += 1
            if val_steps_per_epoch == validation_steps:
                break
        else:
            break
    val_steps_per_epoch = 0
    model.save_weights()
```

#I tried to debug the memory leak without resetting the default graph but no success. It would be very nice if we can debug the problem.

**Comment by Saduf2019** [comment]
@
code shared is incomplete for us to replicate the issue faced, please find [gist](https://colab.sandbox.google.com/gist/Saduf2019/898af8475ecd8114e1692647901a921f/38002.ipynb) of error faced.

**Comment by Saduf2019** [comment]
@cranedroesch
please update as per above comment

**Comment by Saduf2019** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
